src/Monitoring/StatsiteMetricsAgent.php

```diff
@@ -0,0 +1,135 @@
+<?php
+/**
+ * @author    Craig Gosman <craig@ingenerator.com>
+ * @licence   BSD-3-Clause
+ */
+
+
+namespace Ingenerator\PHPUtils\Monitoring;
+
+use DateTimeImmutable;
+use Ingenerator\PHPUtils\DateTime\DateTimeDiff;
+use function array_shift;
+use function explode;
+use function gethostname;
+use function socket_sendto;
+use function sprintf;
+use function strlen;
+
+/**
+ * Supports:
+ *  - COUNTER
+ *  - GAUGE
+ *  - KEY VALUE
+ *  - TIMER
+ *
+ * See https://github.com/statsite/statsite
+ * for more info on metric types supported by statsite
+ *
+ */
+class StatsiteMetricsAgent implements MetricsAgent
+{
+    private const TYPE_COUNTER  = 'c';
+    private const TYPE_GAUGE    = 'g';
+    private const TYPE_KEYVALUE = 'kv';
+    private const TYPE_TIMER    = 'ms';
+
+    /**
+     * @var false|resource
+     */
+    protected $socket;
+
+    protected string $statsite_host;
+
+    protected int $statsite_port;
+
+    protected string $source_hostname;
+
+    public function __construct(string $statsite_host = '127.0.0.1', int $statsite_port = 8125)
+    {
+        $this->statsite_host = $statsite_host;
+        $this->statsite_port = $statsite_port;
+        $this->socket        = socket_create(AF_INET, SOCK_DGRAM, SOL_UDP);
+
+        $this->setSourceHostname(gethostname());
+    }
+
+    public function __destruct()
+    {
+        if ($this->socket) {
+            socket_close($this->socket);
+        }
+    }
+
+    /**
+     * Traditional timer metric calculated stored as floating point in milliseconds
```

Comment:
Probably leave off docblocks here (or just have `@inheritdoc` tags) so they don't drift from interface definitions over time.

Reply:
Most of those comments relate to the specifics of the implementation with statsite ie how we translate our interface to theirs